### PR TITLE
Validate non negative numbers

### PR DESCRIPTION
**Pull request by presidento**
Add rules to grammar to be able to require non-negative and positive integers. Some of our configuration options expected those kind of values, but they were not validated before. (Just the program crashed due an invalid setting.)

(@MrAnno collected this list before, it was reviewed and splitted into the positive and the non-negative parts. Copied from https://github.com/balabit/syslog-ng/pull/1330#issuecomment-275225263)

## Positive numbers

* [listen-backlog()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-syslog-chapter.html#idp4343616)
* [log-fetch-limit()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-syslog-chapter.html#idp4358576)
* [log-iw-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-syslog-chapter.html#idp4368480)
* [max-connections()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-unixstream.html#idp5303904)
* [retries()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-stomp.html#idp11689536)
* [log-msg-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-log-msg-size)
* [time-reap()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-time-reap)
* [time-reopen()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-time-reopen)
* [flush-lines()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-flush-lines)
* [flush-timeout()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-flush-timeout)
* stats-lifetime()
* [dns-cache-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-dns-cache-size)
* [dns-cache-expire-failed()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-dns-cache-expire-failed)
* [port(), destport()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-network-chapter.html#idp9307600) - except where string is also accepted.
* [max-field-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-journal.html#idp4900032)
* [log-fifo-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-log-fifo-size)
* [dns-cache-expire()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-dns-cache-expired)

## Non-negative numbers

* [follow-freq()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-pipe.html#idp3350624)
* [mark-freq()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-pipe.html#idp3375392)
* [so-rcvbuf()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-network.html#idp3095424)
* [so-sndbuf()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-network.html#idp3111616)
* [time-sleep()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#idp14247072)
* [pad-size() in source](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-source-file.html#idp2586352)
* [pad-size() in file destination](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-file.html#idp7278416)
* ip-ttl()
* ip-tos()
* [throttle()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-amqp.html#idp5714800)
* [frac-digits()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-frac-digits)
* [overwrite-if-older()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-file.html#idp7254400)
* [tcp-keepalive-intvl()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-network-chapter.html#idp9373040)
* [tcp-keepalive-probes()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-network-chapter.html#idp9398144)
* [tcp-keepalive-time()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-network-chapter.html#idp9423200)
* suppress()
* [stats-freq()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-options.html#global-option-stats-freq)
* [localport()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-network-chapter.html#idp9186944)
* [disk-buffer / mem-buf-length()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-http-nonjava.html#idp8012832)
* [disk-buffer / mem-buf-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-http-nonjava.html#idp8012832)
* [disk-buffer / disk-buf-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-http-nonjava.html#idp8012832)
* [disk-buffer / qout-size()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-http-nonjava.html#idp8012832)
* [sql retries()](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-sql.html#sql-option-retry-sql-inserts)
* [timeout() (riemann)](https://www.balabit.com/documents/syslog-ng-ose-3.9-guides/en/syslog-ng-ose-guide-admin/html/reference-destination-riemann.html#idp10610416)

## Not changed

* ip-protocol() - only 4 and 6 is accepted
* concurrent_requests() - handled differently, not in grammar
* stats-level() - only 0-3 values are accepted
* dir-perm() - more validation would be needed
* perm() - more validation would be needed
* disk-buf-size() - special values are allowed, documented

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build ERROR

**Comment by kira-syslogng** [comment]
Build ERROR

File: lib/cfg-grammar.y [hunk 1]
```diff
@@ -924,15 +924,15 @@ options_item
...
+	| KW_STATS_LEVEL '(' nonnegative_integer ')'         { last_stats_options->level = $3; }
+	| KW_STATS_LIFETIME '(' positive_integer ')'      { last_stats_options->lifetime = $3; }
 	;
 
 dns_cache_option
-	: KW_DNS_CACHE_SIZE '(' LL_NUMBER ')'	{ last_dns_cache_options->cache_size = $3; }
-	| KW_DNS_CACHE_EXPIRE '(' LL_NUMBER ')'	{ last_dns_cache_options->expire = $3; }
-	| KW_DNS_CACHE_EXPIRE_FAILED '(' LL_NUMBER ')'
+	: KW_DNS_CACHE_SIZE '(' positive_integer ')'	{ last_dns_cache_options->cache_size = $3; }
+	| KW_DNS_CACHE_EXPIRE '(' nonnegative_integer ')'	{ last_dns_cache_options->expire = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: I think this is semantically incorrect. Is there any reason why you chose `nonnegative_integer`?

We have the following logic in the `dns_cache_lookup` function:
```
      if (entry->resolved &&
          ((entry->positive && entry->resolved < now - self->options->expire) ||
           (!entry->positive && entry->resolved < now - self->options->expire_failed)))
        {
          /* the entry is not persistent and is too old */
        }
```

File: lib/cfg-grammar.y [hunk 2]
```diff
@@ -924,15 +924,15 @@ options_item
...
+	| KW_DNS_CACHE_EXPIRE_FAILED '(' positive_integer ')'
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 3]
```diff
@@ -1029,7 +1029,7 @@ parser_opt
 /* LogSource related options */
 source_option
         /* NOTE: plugins need to set "last_source_options" in order to incorporate this rule in their grammar */
-	: KW_LOG_IW_SIZE '(' LL_NUMBER ')'	{ last_source_options->init_window_size = $3; }
+	: KW_LOG_IW_SIZE '(' positive_integer ')'	{ last_source_options->init_window_size = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 4]
```diff
@@ -1049,7 +1049,7 @@ source_proto_option
                         "unknown encoding %s", $3);
             free($3);
           }
-	| KW_LOG_MSG_SIZE '(' LL_NUMBER ')'	{ last_proto_server_options->max_msg_size = $3; }
+	| KW_LOG_MSG_SIZE '(' positive_integer ')'	{ last_proto_server_options->max_msg_size = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 5]
```diff
@@ -1086,7 +1086,7 @@ source_reader_option
 
 	: KW_CHECK_HOSTNAME '(' yesno ')'	{ last_reader_options->check_hostname = $3; }
 	| KW_FLAGS '(' source_reader_option_flags ')'
-	| KW_LOG_FETCH_LIMIT '(' LL_NUMBER ')'	{ last_reader_options->fetch_limit = $3; }
+	| KW_LOG_FETCH_LIMIT '(' positive_integer ')'	{ last_reader_options->fetch_limit = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 6]
```diff
@@ -1104,16 +1104,16 @@ driver_option
     ;
 
 threaded_dest_driver_option
-	: KW_RETRIES '(' LL_NUMBER ')'
+	: KW_RETRIES '(' positive_integer ')'
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: This should be `nonnegative_integer` based on this code snippet in logthrdestdrv.c:

```
 if (self->retries.max <= 0)
    {
      msg_warning("Wrong value for retries(), setting to default",
                ...
    }
```

When you try to send the message for the first time, it won't be counted as a "try", so you can specify 0 here.
:-1:

File: lib/logthrdestdrv.c [hunk 7]
```diff
@@ -327,15 +327,6 @@ log_threaded_dest_driver_start(LogPipe *s)
       return FALSE;
     }
 
-  if (self->retries.max <= 0)
```
Comment (MrAnno) [inline lib/logthrdestdrv.c]: :+1:

File: lib/cfg-grammar.y [hunk 8]
```diff
@@ -1104,16 +1104,16 @@ driver_option
...
         {
           log_threaded_dest_driver_set_max_retries(last_driver, $3);
         }
 
 dest_driver_option
         /* NOTE: plugins need to set "last_driver" in order to incorporate this rule in their grammar */
 
-	: KW_LOG_FIFO_SIZE '(' LL_NUMBER ')'	{ ((LogDestDriver *) last_driver)->log_fifo_size = $3; }
-	| KW_THROTTLE '(' LL_NUMBER ')'         { ((LogDestDriver *) last_driver)->throttle = $3; }
+	: KW_LOG_FIFO_SIZE '(' nonnegative_integer ')'	{ ((LogDestDriver *) last_driver)->log_fifo_size = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: This should be `positive_integer`.
Specifying this option to 0, all of your messages will be dropped:

```
 Destination queue full, dropping messages; queue_len='0'
```

:-1:

File: lib/cfg-grammar.y [hunk 9]
```diff
@@ -1104,16 +1104,16 @@ driver_option
...
+	| KW_THROTTLE '(' nonnegative_integer ')'         { ((LogDestDriver *) last_driver)->throttle = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 10]
```diff
@@ -1144,9 +1144,9 @@ dest_writer_option
         /* NOTE: plugins need to set "last_writer_options" in order to incorporate this rule in their grammar */
 
 	: KW_FLAGS '(' dest_writer_options_flags ')' { last_writer_options->options = $3; }
-	| KW_FLUSH_LINES '(' LL_NUMBER ')'		{ last_writer_options->flush_lines = $3; }
-	| KW_FLUSH_TIMEOUT '(' LL_NUMBER ')'	{ last_writer_options->flush_timeout = $3; }
-        | KW_SUPPRESS '(' LL_NUMBER ')'            { last_writer_options->suppress = $3; }
+	| KW_FLUSH_LINES '(' positive_integer ')'		{ last_writer_options->flush_lines = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: This should be nonnegative_integer, flush_lines(0) means that the output buffering is off.
(I know, this is equivalent to flush_lines(1), but I think 0 is semantically correct as well).
:-1:

File: lib/cfg-grammar.y [hunk 11]
```diff
@@ -1144,9 +1144,9 @@ dest_writer_option
...
+	| KW_FLUSH_TIMEOUT '(' positive_integer ')'	{ last_writer_options->flush_timeout = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 12]
```diff
@@ -1144,9 +1144,9 @@ dest_writer_option
...
+        | KW_SUPPRESS '(' nonnegative_integer ')'            { last_writer_options->suppress = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 13]
```diff
@@ -1155,8 +1155,8 @@ dest_writer_option
 	                                          free($3);
 	                                        }
 	| KW_TEMPLATE_ESCAPE '(' yesno ')'	{ log_writer_options_set_template_escape(last_writer_options, $3); }
-	| KW_PAD_SIZE '(' LL_NUMBER ')'         { last_writer_options->padding = $3; }
-	| KW_MARK_FREQ '(' LL_NUMBER ')'        { last_writer_options->mark_freq = $3; }
+	| KW_PAD_SIZE '(' nonnegative_integer ')'         { last_writer_options->padding = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: lib/cfg-grammar.y [hunk 14]
```diff
@@ -1155,8 +1155,8 @@ dest_writer_option
...
+	| KW_MARK_FREQ '(' nonnegative_integer ')'        { last_writer_options->mark_freq = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: `0 means "no MARK message"`
:+1:

File: lib/cfg-grammar.y [hunk 15]
```diff
@@ -1189,7 +1189,7 @@ file_perm_option
 
 template_option
 	: KW_TS_FORMAT '(' string ')'		{ last_template_options->ts_format = cfg_ts_format_value($3); free($3); }
-	| KW_FRAC_DIGITS '(' LL_NUMBER ')'	{ last_template_options->frac_digits = $3; }
+	| KW_FRAC_DIGITS '(' nonnegative_integer ')'	{ last_template_options->frac_digits = $3; }
```
Comment (MrAnno) [inline lib/cfg-grammar.y]: :+1:

File: modules/afsocket/afsocket-grammar.ym [hunk 16]
```diff
@@ -717,28 +713,20 @@ tls_option
...
-	{
-		CHECK_ERROR($3 >= 0 && $3 <= G_MAXINT, @3, "Invalid so_sndbuf, it has to be between 0 and %d", G_MAXINT);
-		last_sock_options->so_sndbuf = $3;
-	}
-	| KW_SO_RCVBUF '(' LL_NUMBER ')'
-	{
-		CHECK_ERROR($3 >= 0 && $3 <= G_MAXINT, @3, "Invalid so_rcvbuf, it has to be between 0 and %d", G_MAXINT);
-		last_sock_options->so_rcvbuf = $3;
-	}
+	: KW_SO_SNDBUF '(' nonnegative_integer ')' { last_sock_options->so_sndbuf = $3; }
```
Comment (gaborznagy) [inline modules/afsocket/afsocket-grammar.ym]: :+1:

File: modules/afsocket/afsocket-grammar.ym [hunk 17]
```diff
@@ -717,28 +713,20 @@ tls_option
...
+	| KW_SO_RCVBUF '(' nonnegative_integer ')' { last_sock_options->so_rcvbuf = $3; }
```
Comment (gaborznagy) [inline modules/afsocket/afsocket-grammar.ym]: :+1:

File: lib/cfg-grammar.y (hunk 6 above)
Comment (presidento) [inline lib/cfg-grammar.y]: This (deleted) code snippet says explicitly that this config option must be a positive number. If it gets a non-positive number, the default value (3) will be applied with a warning message.

File: lib/cfg-grammar.y [hunk 18]
```diff
@@ -884,32 +884,32 @@ options_items
...
+	| KW_SUPPRESS '(' nonnegative_integer ')'		{ configuration->suppress = $3; }
 	| KW_THREADED '(' yesno ')'		{ configuration->threaded = $3; }
 	| KW_PASS_UNIX_CREDENTIALS '(' yesno ')' { configuration->pass_unix_credentials = $3; }
 	| KW_USE_RCPTID '(' yesno ')'		{ cfg_set_use_uniqid($3); }
 	| KW_USE_UNIQID '(' yesno ')'		{ cfg_set_use_uniqid($3); }
-	| KW_LOG_FIFO_SIZE '(' LL_NUMBER ')'	{ configuration->log_fifo_size = $3; }
-	| KW_LOG_IW_SIZE '(' LL_NUMBER ')'	{ msg_warning("WARNING: Support for the global log-iw-size() option was removed, please use a per-source log-iw-size()", cfg_lexer_format_location_tag(lexer, &@1)); }
-	| KW_LOG_FETCH_LIMIT '(' LL_NUMBER ')'	{ msg_warning("WARNING: Support for the global log-fetch-limit() option was removed, please use a per-source log-fetch-limit()", cfg_lexer_format_location_tag(lexer, &@1)); }
-	| KW_LOG_MSG_SIZE '(' LL_NUMBER ')'	{ configuration->log_msg_size = $3; }
+	| KW_LOG_FIFO_SIZE '(' nonnegative_integer ')'	{ configuration->log_fifo_size = $3; }
```
Comment (presidento) [inline lib/cfg-grammar.y]: Done.

File: lib/cfg-grammar.y (hunk 1 above)
Comment (presidento) [inline lib/cfg-grammar.y]: Done.

File: lib/cfg-grammar.y (hunk 8 above)
Comment (presidento) [inline lib/cfg-grammar.y]: Done.

**Comment by kira-syslogng** [comment]
Build ERROR

File: lib/cfg-grammar.y (hunk 6 above)
Comment (MrAnno) [inline lib/cfg-grammar.y]: Right, sorry. :+1:

**Comment by MrAnno** [comment]
The error message of the invalid `so-rcvbuf()` and `so-sndbuf()` options have been changed. Kira's tests should be modified too.

File: lib/cfg-grammar.y [hunk 19]
```diff
@@ -881,32 +884,32 @@ options_items
 	;
 
 options_item
-	: KW_MARK_FREQ '(' LL_NUMBER ')'		{ configuration->mark_freq = $3; }
-	| KW_FLUSH_LINES '(' LL_NUMBER ')'		{ configuration->flush_lines = $3; }
+	: KW_MARK_FREQ '(' nonnegative_integer ')'		{ configuration->mark_freq = $3; }
+	| KW_FLUSH_LINES '(' positive_integer ')'		{ configuration->flush_lines = $3; }
```
Comment (kvch) [inline lib/cfg-grammar.y]: AFAIK 0 is a valid value for `flush-lines`. It means that the buffering is off.

Comment (kvch) [inline lib/cfg-grammar.y]: You are right.

Comment (kvch) [inline lib/cfg-grammar.y]: Yes, you are right.

**Comment by presidento** [comment]
@pzoleex can you change the test cases of @kira-syslogng? It is failing at the moment, because we initiated an universal rule and error message for the non-negative numbers, but there are tests for validating `so_rcvbuf`. After that this can be merged, as @kvch and @MrAnno said.

**Comment by presidento** [comment]
@kira-syslogng retest this please

**Comment by pzoleex** [comment]
I fixed the testcase now, but there is still a problem:
if the rcv_buf is higher then 2147483647 (2 GB) syslog-ng should not start because the system handles it as a negative number (overflowing).
Without this patch, syslog-ng stops and gives an error message:
Error parsing afsocket, Invalid so_rcvbuf, it has to be between 0 and 2147483647

after this patch, syslog-ng starts (with a wrong value).
Basically this is the config:
source s_network_2850b063c3ca465b99b864d46a91e847 {
    network(ip("0.0.0.0")
 ip_protocol(4)
 log_iw_size(100)
 max_connections(1)
 port(10514)
 so_rcvbuf(2147483648));
};

the test fixes are on this branch:
pzolee-validate-non-negative-numbers
but I will only integrate it when the above error is fixed

**Comment by presidento** [comment]
@pzoleex I have fixet it, it sais

```
Error parsing afsocket, Invalid so_rcvbuf, it has to be less than 
2147483647 in etc/syslog-ng.conf at line 9, column 15:

    so_rcvbuf(2147483648));
              ^^^^^^^^^^
```

It starts with the maximum allowed number, but there is a warning:

```
[2017-03-24T14:12:36.038363] The kernel refused to set the receive buffer
(SO_RCVBUF) to the requested size, you probably need to adjust buffer
related kernel parameters; so_rcvbuf='2147483647', so_rcvbuf_set='425984'

```

I think it is OK, and this is why I have written _less than_ and not _less or equal than_. ;)

Is it OK this way?

**Comment by pzoleex** [comment]
ok

**Comment by kira-syslogng** [comment]
Build ERROR

**Comment by kira-syslogng** [comment]
Build ERROR

**Comment by presidento** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS
